Выполнить работу спринта №6.

SergeyLeshchev opened the pull request:
1. Сделать историю неограниченной по количеству задач.
2. Сделать историю без повторяющихся задач.
3. Добавить метод remove в HistoryManager.
4. Добавить класс Node.
5. Создать реализацию двусвязного списка.
6. Добавить методы linkLast, removeNode и getTasks.
7. Добавить необходимые unit-тесты.
8. Выполнить дополнительное задание.

Commit:
Выполнить работу спринта №6.

1. Сделать историю неограниченной по количеству задач.
2. Сделать историю без повторяющихся задач.
3. Добавить метод remove в HistoryManager.
4. Добавить класс Node.
5. Создать реализацию двусвязного списка.
6. Добавить методы linkLast, removeNode и getTasks.
7. Добавить необходимые unit-тесты.
8. Выполнить дополнительное задание.

Commit:
Исправить ошибки в кодстайл

Commit:
Исправить ошибки в кодстайл

Ivanov-s-tmn commented on src/data/Node.java:
Класс ноды можно сделать вложенным в InMemoryHistoryManager (private static), он не нужен как отдельный класс, потому что используется для реализации истории внутри менеджера истории. Соответственно и не нужны геттеры и сеттеры для доступа к полям, так же как и не нужны доп методы equals() и hashCode(), все сравнение будет происходить по ссылкам

Плюс, не нужна и параметризация <T>, в нашем случае это только усложняет синтаксис, мы всегда знаем, что в нашей истории будет храниться тип `Task `

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -19,12 +20,44 @@ void addTaskInHistoryAndGetHistoryTest() {
-        for (int i = 0; i < 20; i++) {
-            historyManager.addTaskInHistory(task);
+        int capacity = 1000000;
+        for (int i = 0; i < capacity; i++) {
+            historyManager.addTaskInHistory(new Task("Title" + i, "Description" + i, Status.NEW, i));
```
В тестах стоит избегать циклов, особенно с таким количество элементов, если конечно не стоит задача протестировать производительность, но это уже за рамками юнит тестов. Вообще, будет достаточно 3 тестовых задач для всех тестов. Можно создать конкретные объекты в `BeforeEach` методе без цикла, добавить их в историю, а в самих тестах достаточно 2 строк кода, как пример:
```java
historyManager.remove(1);
assertEquals(List.of(task2, task3), historyManager.getHistoryList());
```
Это пример для теста удаления задачи из начала истории, раз сейчас у нас реализован связный список, в котором может быть удаления из разных частей истории (из начала, из середины и с конца), нам важно это покрыть тестами. Можно взять за основу предложенный выше код.

SergeyLeshchev replied:
в этом задании мы меняли историю так, чтобы она могла хранить бесконечное количество элементов. Чтобы проверить новую возможность программы я добавил тест на вместимость истории. Способа кроме добавления в историю миллиона задач я не придумал. Я бы добавил больше, если бы был комп помощнее)) тест на 1 млн задач проходит менее чем за 3 секунды, но тестов может быть очень много, я понимаю, что создание длительных тестов ни есть хорошо. Возможно этот тест совсем не нужен, так как очевидно, что история может быть очень большой. Как я понял, то размер ограничен только размером стандартной HashMap, которая содержит ключ и Node и может быть размером числа максимального int/2

SergeyLeshchev replied:
Создал метод @BeforeEach в тестах, но так как больше 2 задач нам нужно только в одном тесте, то в методе генерируется каждый раз 2 задачи. А третья и четвертая по необходимости внутри теста

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
 
-    private final int CAPACITY_HISTORY = 10;
-    private final ArrayList<Task> history = new ArrayList<>();
+    private final Map<Integer, Node<Task>> mapForHistory = new HashMap<>();
+    private final LinkedListForHistory<Task> linkedListForHistory = new LinkedListForHistory<>();
```
Если принято решение выделить отдельный класс `LinkedListForHistory` для реализации связного списка, то тогда нужно спрятать хешмапу в него к нодам, ведь на основе нее и нод last first он и реализован

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
         }
-        if (history.size() == CAPACITY_HISTORY) {
-            history.removeFirst();
+        if (mapForHistory.containsKey(task.getIdOfTask())) {
```
Проверка лишняя, есть проверка `task == null` в начале этого метода, и есть проверка `node == null` в методе `removeNode()` и нам нечего больше бояться)

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
+        mapForHistory.remove(id);
+    }
+
+    private ArrayList<Task> getTasks() {
+        ArrayList<Task> history = new ArrayList<>();
+        Node<Task> node = linkedListForHistory.first;
+        while (history.size() != linkedListForHistory.size) {
+            history.add(node.getTask());
+            node = node.getNext();
+        }
```
Этот метод тоже стоит спрятать в `LinkedListForHistory` к другим методам с логикой, а метод `getHistory()` будет его вызывать

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
-    public ArrayList<Task> getHistory() {
+    public void remove(int id) {
+        linkedListForHistory.removeNode(mapForHistory.get(id));
+        mapForHistory.remove(id);
```
Удаление из хешмапы и добавление из хешмапы тоже соответственно нужно будет перенести в логику методов внутри `LinkedListForHistory` после переноса хешмапы

SergeyLeshchev replied:
у меня нет собственной реализации логики методов по добавлению в хешмапу и удалению из хешмапы. Это делают методы из библиотеки. Но так как сама хешмапа теперь в классе LinkedListForHistory, то я вызвал методы через обращение к хешмапе, которая является полем класса LinkedListForHistory. Если имелось ввиду не это, то ответь пожалуйста как сделать будет правильней. Я так сделал потому что в задании указано, что метод remove должен быть прописан в интерфейсе HistoryManager. Кроме того если перенести метод remove в LinkedListForHistory, то во всех вызовах remove будет более длинное обращение, что я думаю не очень удобно и красиво

Ivanov-s-tmn replied:
Просто перенести удаление из хешмапы в логику `removeNode()` и все:
```java
private void removeNode(Node node) {
            if (node == null) {
                return;
            }
            if (node == first && node == last) {
                first = null;
                last = null;
            } else if (node.next == null) { // Если node - хвост
                node.prev.next = null;
                last = node.prev;
            } else if (node.prev == null) { // Если node - голова
                node.next.prev = null;
                first = node.next;
            } else {
                node.next.prev = node.prev;
                node.prev.next = node.next;
            }
            mapForHistory.remove(node.task.getId()); //удаление из хешмапы
}
```
А сейчас получается, у тебя в методе добавления новой ноды в конец нет удаления из хешмапы, так как там вызывается метод `removeNode()` а в нем не заложено удаление
А так тебе будет достаточно вызывать метод `removeNode()` из методов добавления и удаления

SergeyLeshchev replied:
Спасибо, понял, исправил

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
+            if (node == null) {
+                return;
+            }
+            if (size == 1) {
```
Можно заменить на проверку `node == first && node == last`, тогда можно будет избавиться от переменной длины списка

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
+
+        private Node<T> linkLast(T task) {
+            Node<T> newNode = new Node<>(task);
+            if (size > 0) {
```
Тут тоже можно заменить на проверку `first != null`, а можно и инвертировать условие на `first == null`

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,86 @@
+        private Node<T> linkLast(T task) {
+            Node<T> newNode = new Node<>(task);
+            if (size > 0) {
+                newNode.setPrev(last);
```
Можно задать ссылку на предыдущую последнюю ноду сразу при создании как пример:
`Node newNode = new Node<>(task, last, null);`

Commit:
Выполнить работу над ошибками в ТЗ спринта №6

1. Перенести Node в LinkedListForHistory
2. Сделать Node не параметризированным
3. Удалить ненужные методы из Node
4. Перенести mapForHistory в LinkedListForHistory
5. Перенести getTasks в LinkedListForHistory
6. Удалить поле size в LinkedListForHistory
7. Удалить ненужную проверку в addTaskInHistory
8. Убрать циклы из тестов и добавить @BeforeEach

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
Я в комментарии к прошлому ревью просил реализовать тесты удаления из разных частей истории (из начала, из середины и с конца), это важно покрыть тестами, потому что у нас сейчас реализован связный список и он сильно зависим от порядка

Commit:
Выполнить работу №2 над ошибками в ТЗ спринта №6

1. Перенести удаление из mapForHistory в метод removeNode
2. Исправить тесты на удаление задач из истории
3. Перенести добавление тестовых задач в историю в BeforeEach
